usort/types.py

```diff
@@ -32,43 +56,106 @@ class SortKey:
-    # os` and `import os.path` do not shadow becuase it's the same `os`
-    imported_names: Dict[str, str] = field(default_factory=dict, compare=False)
+@dataclass(order=True, repr=False)
+class SortableImport:
+    sort_key: SortKey = field(init=False)
+    stem: Optional[str] = field(order=case_insensitive_ordering)  # "from" imports
+    items: Sequence[SortableImportItem] = field()
+    comments: ImportComments = field()
+
+    config: Config = field(order=False, factory=Config)
+    node: cst.CSTNode = field(factory=cst.EmptyLine)  # for cli/debugging only
+
+    def __repr__(self) -> str:  # pragma: nocover
+        items = indent(("\n".join(f"{item!r}," for item in self.items)), "        ")
+        return (
+            dedent(
+                """
+            SortableImport(
+                # sort_key = {sort_key!r},
+                stem = {stem!r},
+                items = [
+            {items}
+                ],
+                comments = {comments!r},
+            )
+            """
+            )
+            .strip()
+            .format(
+                sort_key=self.sort_key,
+                stem=self.stem,
+                items=items,
+                comments=self.comments,
+            )
+        )
 
-    def __post_init__(self) -> None:
-        if not self.first_module.startswith("."):
-            ndots = 0
+    @property
+    def imported_names(self) -> Dict[str, str]:
+        results: Dict[str, str] = {}
+
+        for item in self.items:
+            key = item.asname or item.name
+            if self.stem is None and not item.asname:
+                key = top_level_name(item.name)
+                value = key
+            else:
+                value = stem_join(stem=self.stem, name=item.name)
+            results[key] = value
+
+        return results
+
+    def __attrs_post_init__(self) -> None:
+        top: Optional[str] = None
+        ndots = 0
+
+        if self.stem is None:
+            top = top_level_name(self.items[0].name)
+        elif not self.stem.startswith("."):
+            top = top_level_name(self.stem)
         else:
             # replicate ... sorting before .. before ., but after absolute
-            ndots = 100 - (len(self.first_module) - len(self.first_module.lstrip(".")))
+            ndots = 100 - (len(self.stem) - len(self.stem.lstrip(".")))
+
+        category = self.config.category(top) if top else CAT_FIRST_PARTY
+
         self.sort_key = SortKey(
             # TODO this will raise on missing category
-            category_index=self.config.categories.index(
-                self.config.category(self.first_module)
-            ),
-            is_from_import=isinstance(self.node.body[0], cst.ImportFrom),
+            category_index=self.config.categories.index(category),
+            is_from_import=bool(self.stem),
             ndots=ndots,
-            module=self.first_module.casefold(),
+            module="",
```

Comment:
I don't see `top` being used, did you mean to say `module=top` here?

Reply:
I don't think we actually need this at all anymore, because the stem is the next thing it will sort on, and then imported items beyond that if that doesn't break the tie. Am I missing something?